connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertHeader.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.transforms;
+
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.connect.connector.ConnectRecord;
+import org.apache.kafka.connect.data.SchemaAndValue;
+import org.apache.kafka.connect.data.Values;
+import org.apache.kafka.connect.transforms.util.SimpleConfig;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class InsertHeader<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    public static final String OVERVIEW_DOC =
+        "Insert header as a literal value.";
+
+    private interface ConfigName {
+        String HEADER_CONFIG = "header";
+        String HEADER_VALUE_CONFIG = "literal.value";
+    }
+
+    /**
+     * Maps known logical types to a list of Java classes that can be used to represent them.
+     */
+    private static final Map<String, List<Class>> LOGICAL_TYPE_CLASSES = new HashMap<>();
+
+    public static final ConfigDef CONFIG_DEF = new ConfigDef()
+        .define(ConfigName.HEADER_CONFIG, ConfigDef.Type.STRING, ConfigDef.NO_DEFAULT_VALUE, ConfigDef.Importance.MEDIUM,
+            "Name of the header to add.")
+        .define(ConfigName.HEADER_VALUE_CONFIG, ConfigDef.Type.STRING, ConfigDef.NO_DEFAULT_VALUE, ConfigDef.Importance.MEDIUM,
+            "Value of the header to add.");
+
+    private String headerName;
+    private String headerValue;
+
+    @Override
+    public void configure(Map<String, ?> props) {
+        final SimpleConfig config = new SimpleConfig(CONFIG_DEF, props);
+        headerName = config.getString(ConfigName.HEADER_CONFIG);
+        headerValue = config.getString(ConfigName.HEADER_VALUE_CONFIG);
+    }
+
+    @Override
+    public R apply(R record) {
+        if (headerValue == null) return null;
+        SchemaAndValue schemaAndValue = Values.parseString(headerValue);
```

Comment:
`SchemaAndValue` is an immutable object and can be reused, so let's move this line into the `configure(...)` method and store it as a final field.

In fact, `ConnectHeader` is immutable, which means this could do the following:
```
    private final ConnectHeader newHeader;
    ...
    public void configure(Map<String, ?> props) {
        ...
        newHeader = new ConnectHeader(headerName, Values.parseString(headerValue));
    }
    ...
    public R apply(R record) {
        if (record == null) {
            return record;
        }
        Headers recordHeaders = record.headers();
        Headers newHeaders = new ConnectHeaders(record.headers());
        newHeaders.add(newHeader);
        return record.newRecord(record.topic(), record.kafkaPartition(), record.keySchema(), record.key(), record.valueSchema(), updatedValue, record.timestamp(), newHeaders);
    }
```
Note how we always create a new record, and don't modify the existing record. This is not the most efficient way to work, but it guarantees the SMT contract.

Reply:
One more thing, KIP-145 doesn't really distinguish for this SMT whether the header should be *added* if it does not already exist, or whether if an existing header with this value exists the SMT should *replace* that value. (Like HTTP, you can have multiple headers with the same key.)

I think it's okay for this SMT to just always add it, since the record likely won't have it to begin with (most source connectors don't add headers, and many sink connectors don't use them).

Reply:
I tried to implement your suggestion but ConnectHeader is not a public class so I cannot instantiate it outside the package.